src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs

```diff
@@ -685,367 +685,221 @@ private static bool TryCopyTo(string source, Span<char> destination, out int cha
             return false;
         }
 
-        public static unsafe string FormatInt32(int value, ReadOnlySpan<char> format, IFormatProvider? provider)
+        /// <summary>
+        /// This interface lets us use shared implementations (FormatInt and TryFormatInt)
```

Comment:
It is true the code in the steady state is going to be almost identical. There is extra overhead in the size of runtime generics control structures. My back of the envelope calculation says that it is 10kB - 20kB (once all integer formatting routines are used).

@marek-safar How do you feel about the trade-offs in this implementation for mobile and web-assembly (footprint-sensitive targets)?

The technique that can be used to eliminate the generics overhead, while keeping the code shared, is generating the code using .tt template. We use it for SIMD types: https://github.com/dotnet/runtime/tree/master/src/libraries/System.Private.CoreLib/src/System/Numerics . It does not look as pretty, but gets the job done. Thoughts about whether it would be worth doing it in this case?

Reply:
The formating code is already casing troubles when linking, introducing more interface abstraction will make the matter worse. It's not a large amount of code but it's quite likely we'll need to redo the formating code to be linker friendly.

Although the numbers look good for JIT we should not overoptimize for JIT and make e.g. AOT slower. What is the perf impact on AOT?

Reply:
> The formating code is already causing troubles when linking

Can you share details?

Reply:
I don't have precise data at hand as we are still working on the tooling but the fact is that the formatting code is a large chunk of any app (minimal amount code is linked out for formatting)

Reply:
Thanks. I thought you were saying it's worse on core than it was previously on mono; that's not the case, right?

Reply:
I don't have the exact data to confirm that but my guess would be that it should not differ significantly between mono and netcore (in a negative or positive way) at the moment.

Reply:
System.Private.CoreLib.dll R2R image size is 9,203,200 before this change and 9,213,440 after this change (Win x64 release). This size delta is not the entire overhead - there is also extra private memory at runtime. It would be nice to reduce it.